File: pyBAScloudAPI/tests/connector_tests.py
import unittest
import time 
import datetime

from definitions import *

import pyBAScloudAPI as api

BCAPI = api.EntityContext(BASCLOUD_TEST_URL)


def errorHandler(e, j):
    raise e

def authenticate():
    currentDateTime = int(time.time())
    print("CURRENT DATETIME: ", currentDateTime)

    currentDateTimeString = datetime.datetime.fromtimestamp(currentDateTime).strftime("%FT%T.000Z")
    print("CURRENT DATETIME STR: ", currentDateTimeString)

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_EMAIL, BASCLOUD_TEST_PASS)

    print("TOKEN: ", BCAPI.getToken())
    print("EXPIRATION DATE: ", datetime.datetime.fromtimestamp(BCAPI.getTokenExpirationDate()).strftime("%FT%T.000Z"))


class TestConnectorCollection(unittest.TestCase):

    def test_connector(self):
        print("\tRequesting all connectors...")

        connectors = BCAPI.getConnectorsCollection(BASCLOUD_TEST_TENANT_UUID)
        print("\t\tOK.")

        print("\tFound properties: ", len(connectors[0]))

        self.assertTrue(len(connectors[0]) >= 1)


class TestSingleConnector(unittest.TestCase):

    def test_connector(self):
        print("\tRequesting single connector with UUID...")

        connector = BCAPI.getConnector(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID)
        print("\t\tOK.")

        print("\tConnector UUID: ", connector.uuid)
        print("\tConnector Name: ", connector.name)
        print("\tConnector created at: ", datetime.datetime.fromtimestamp(connector.createdAt).strftime("%FT%T.000Z"))
        print("\tConnector updated at: ", datetime.datetime.fromtimestamp(connector.updatedAt).strftime("%FT%T.000Z"))

        self.assertEqual(connector.uuid, BASCLOUD_TEST_CONNECTOR_UUID)
        self.assertEqual(connector.name, "Stefan01")
        self.assertTrue(connector.createdAt > 0)
        self.assertTrue(connector.updatedAt > 0)


class TestAssociatedProperty(unittest.TestCase):

    def test_connector(self):
        print("\tRequesting associated property of the connector again...")

        conn_prop = BCAPI.getAssociatedProperty(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID)
        print("\t\tOK.")

        print("\tConnector's property UUID: ", conn_prop.uuid)

        self.assertEqual(conn_prop.uuid, BASCLOUD_TEST_PROPERTY_UUID)


class TestAssociatedDevices(unittest.TestCase):

    def test_connector(self):
        print("\tRequest connector's associated devices...")

        conn_devices = BCAPI.getAssociatedDevices(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID)

        print("\t\tFound connector with ", len(conn_devices[0]), " devices")

        self.assertTrue(len(conn_devices[0]) >= 1)


class TestCreateUpdateAndDeleteConnector(unittest.TestCase):

    def test_connector(self):
        print("\tCreating new connector...")

        new_connector = BCAPI.createConnector(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_PROPERTY_UUID, "TestConnector")
        print("\t\tOK.")

        new_uuid = new_connector.uuid

        self.assertEqual(new_connector.name, "TestConnector")
        self.assertTrue(new_connector.createdAt > 0)
        self.assertTrue(new_connector.updatedAt > 0)

        print("\tUpdating newly created connector...")

        up_re_connector = BCAPI.updateConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid, "TestConnector2")
        print("\t\tOK.")

        self.assertEqual(up_re_connector.name, "TestConnector2")


        print("\tDeleting newly created connector...")
        BCAPI.deleteConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        print("\t\tOK.")

        try:
            non_connector = BCAPI.getConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid)
        except api.NotFoundRequest:
            self.assertTrue(True)
        except:
            self.assertTrue(False)


if __name__ == '__main__':
    authenticate()
    unittest.main()

File: tests/connector_tests.cpp
#include <chrono>
#include <iostream>

#include <fmt/core.h>
#include <fmt/chrono.h>

#include <gtest/gtest.h>

#include "EntityContext.h"
#include "entity/EntityCollection.h"
#include "error/Exceptions.h"


using namespace BAScloud;

EntityContext BCAPI(BASCLOUD_TEST_URL);


void authenticate() {

    std::time_t currentDateTime = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch()).count();

    std::cerr <<"CURRENT DATETIME: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(currentDateTime))<<std::endl;

    BCAPI.authenticateWithUserLogin(BASCLOUD_TEST_ADMIN_EMAIL, BASCLOUD_TEST_PASS);

    std::cerr <<"TOKEN: "<<BCAPI.getToken()<<std::endl;
    std::cerr <<"EXPIRATION DATE: "<<fmt::format("{:%Y-%m-%d %H:%M:%S}", fmt::localtime(BCAPI.getTokenExpirationDate()))<<std::endl;
}

TEST(BasicTests, ConnectorCollectionTest) {

    std::cout << "\tRequesting all connectors..." << std::endl;

    EntityCollection<Connector> connectors = BCAPI.getConnectorsCollection(BASCLOUD_TEST_TENANT_UUID);
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tFound properties: " << connectors.first.size() << std::endl;

    EXPECT_TRUE(connectors.first.size() >= 1);
}

TEST(BasicTests, SingleConnectorTest) {

    std::cout << "\tRequesting single connector with UUID..." << std::endl;

    Connector connector = BCAPI.getConnector(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID);
    std::cout << "\t\tOK." << std::endl;

    std::cout << "\tConnector UUID: " << connector.getUUID() << std::endl;
    std::cout << "\tConnector Name: " << connector.getName() << std::endl;
    std::cout << "\tConnector created at: " << fmt::format("{:%F %T}", fmt::localtime(connector.getCreatedDate())) << std::endl;
    std::cout << "\tConnector updated at: " << fmt::format("{:%F %T}", fmt::localtime(connector.getLastUpdatedDate())) << std::endl;

    EXPECT_EQ(connector.getUUID(), BASCLOUD_TEST_CONNECTOR_UUID);
    EXPECT_EQ(connector.getName(), "Stefan01");
    EXPECT_TRUE(connector.getCreatedDate() > 0);
    EXPECT_TRUE(connector.getLastUpdatedDate() > 0);
}

// Deprecated
// TEST(BasicTests, AssociatedPropertyTest) {

//     std::cout << "\tRequesting associated property of the connector again..." << std::endl;

//     Property conn_prop = BCAPI.getAssociatedProperty(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID);
//     std::cout << "\t\tOK." << std::endl;

//     std::cout << "\tConnector's property UUID: " << conn_prop.getUUID() << std::endl;

//     EXPECT_EQ(conn_prop.getUUID(), BASCLOUD_TEST_PROPERTY_UUID);
// }

TEST(BasicTests, AssociatedDevicesTest) {

    std::cout << "\tRequest connector's associated devices..." << std::endl;
    PagingOption paging = {};

    EntityCollection<Device> conn_devices = BCAPI.getAssociatedConnectorDevices(BASCLOUD_TEST_TENANT_UUID, BASCLOUD_TEST_CONNECTOR_UUID);

    std::cout <<"\t\tFound connector with " << conn_devices.first.size() << " devices" << std::endl;

    EXPECT_TRUE(conn_devices.first.size() >= 1);
}

TEST(BasicTests, CreateUpdateAndDeleteConnectorTest) {

    std::cout << "\tCreating new connector..." << std::endl;

    Connector new_connector = BCAPI.createConnector(BASCLOUD_TEST_TENANT_UUID, "TestConnector");
    std::cout << "\t\tOK." << std::endl;

    std::string new_uuid = new_connector.getUUID();

    EXPECT_EQ(new_connector.getName(), "TestConnector");
    EXPECT_TRUE(new_connector.getCreatedDate() > 0);
    EXPECT_TRUE(new_connector.getLastUpdatedDate() > 0);

    std::cout << "\tUpdating newly created connector..." << std::endl;

    Connector up_re_connector = BCAPI.updateConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid, "TestConnector2");
    std::cout << "\t\tOK." << std::endl;

    EXPECT_EQ(up_re_connector.getName(), "TestConnector2");


    std::cout << "\tDeleting newly created connector..." << std::endl;
    BCAPI.deleteConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    std::cout << "\t\tOK." << std::endl;

    try {
        Connector non_connector = BCAPI.getConnector(BASCLOUD_TEST_TENANT_UUID, new_uuid);
    } catch(NotFoundRequest& e) {
        EXPECT_TRUE(true);
    } catch(...) {
        EXPECT_TRUE(false);
    }

}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);

    authenticate();

    return RUN_ALL_TESTS();
}